### PR TITLE
PWN-4858 - Restore Logic Device + Custom Share

### PR DESCRIPTION
**Pull request by davran1001**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-4858

## Description of Work

Restore Logic with ThirdShare(SMS restore - GREEN FLOW WORKDS)

### CODE REVIEW

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/GatewayServiceRemoteRepository.kt [hunk 1]
```diff
@@ -33,20 +35,23 @@ class GatewayServiceRemoteRepository(
     private var restoreWalletPublicKey: Base58String? = null
         set(value) {
             field = value
-            resetTemporaryPublicKeyTimer.launchTimer(appScope)
+            if (value != null) {
+                resetTemporaryPublicKeyTimer.launchTimer(appScope)
+                Timber.i("restoreWalletPublic key value = $value")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/GatewayServiceRemoteRepository.kt:40]: forgot to remove?

File: app/src/main/java/org/p2p/wallet/auth/interactor/restore/CustomShareRestoreInteractor.kt [hunk 2]
```diff
@@ -59,6 +62,7 @@ class CustomShareRestoreInteractor(
         thirdShareStructAsBase64: String
     ): Web3AuthSignUpResponse.ShareDetailsWithMeta {
         val thirdShareJson = String(thirdShareStructAsBase64.decodeFromBase64())
+        Timber.tag("___________").d(thirdShareJson)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/restore/CustomShareRestoreInteractor.kt:65]: ______?

File: app/src/main/java/org/p2p/wallet/auth/model/PhoneNumber.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+            return "+$trimmedValue"
+        }
+    }
+}
+
+fun PhoneNumber.e164Formatted(): String {
+    return PhoneNumber.PhoneNumberFormatter.getPhoneNumberE164Formatted(this)
+}
+
+fun PhoneNumber.equals(phoneNumber: PhoneNumber?): Boolean {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/model/PhoneNumber.kt:18]: !!! please override equals inside PhoneNumber, i don't think == != will work with this method 

e164Formatted can be a method too

File: app/src/main/java/org/p2p/wallet/auth/model/PhoneNumber.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package org.p2p.wallet.auth.model
+
+data class PhoneNumber(val value: String) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/model/PhoneNumber.kt:3]: from the name of the class i can't understand how the value looks inside
is it +7 123 456 78 90 or +71234567890 or 71234567890

File: app/src/main/java/org/p2p/wallet/auth/ui/smsinput/NewSmsInputPresenter.kt [hunk 5]
```diff
@@ -132,7 +138,10 @@ class NewSmsInputPresenter(
 
                 view?.renderButtonLoading(isLoading = true)
 
-                createWalletInteractor.startCreatingWallet(repository.userPhoneNumber.orEmpty())
+                createWalletInteractor.startCreatingWallet(
+                    signUpRepository.userPhoneNumber
+                        ?: throw IllegalStateException("User phone number cannot be null")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/ui/smsinput/NewSmsInputPresenter.kt:143]:
```suggestion
                    signUpRepository.userPhoneNumber ?: error("User phone number cannot be null")
```

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/auth/ui/smsinput/NewSmsInputPresenter.kt:143]: and please take userPhoneNumber inside the interactor

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/GatewayServiceErrorMapper.kt [hunk 6]
```diff
@@ -31,6 +31,9 @@ class GatewayServiceErrorMapper {
             -32057 -> GatewayServiceError.UserAlreadyExists(
                 error.errorCode, error.errorMessage
             )
+            -32059 -> GatewayServiceError.TooOftenOtpRequests(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/gateway/repository/GatewayServiceErrorMapper.kt:34]: TooManyOtpRequests

File: app/src/main/java/org/p2p/wallet/auth/interactor/AuthLogoutInteractor.kt [hunk 7]
```diff
@@ -47,18 +46,4 @@ class AuthLogoutInteractor(
             pushNotificationsInteractor.deleteDeviceToken(publicKey)
         }
     }
-
-    fun clearAppData() {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/AuthLogoutInteractor.kt:51]: maybe we should stop the services at least?

File: app/src/main/java/org/p2p/wallet/auth/interactor/restore/CustomShareRestoreInteractor.kt [hunk 8]
```diff
@@ -6,35 +6,39 @@ import org.p2p.solanaj.utils.crypto.decodeFromBase64
...
+                solanaPublicKey = temporaryUserPublicKey,
+                solanaPrivateKey = temporaryUserPrivateKey,
+                phoneNumber = userPhoneNumber,
+                channel = OtpMethod.SMS
+            )
+        }
         restoreFlowDataLocalRepository.userPhoneNumber = userPhoneNumber
     }
 
-    suspend fun finishRestoreCustomShare(smsCode: String) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/restore/CustomShareRestoreInteractor.kt:37]: revert this please, it's CustomShareRestoreInteractor, so it's methods should be called accordingly 
on the domain layer - we restore custom share
on data layer - we use restore / confirm_restore_wallet methods for Gateway

File: app/src/main/java/org/p2p/wallet/auth/interactor/restore/UserRestoreInteractor.kt [hunk 9]
```diff
@@ -42,57 +46,109 @@ class UserRestoreInteractor(
...
+
+    private suspend fun tryRestoreUser(
+        restoreWay: OnboardingFlow.RestoreWallet.SocialPlusCustomShare
+    ): RestoreUserResult = try {
+        val customShare = restoreFlowDataLocalRepository.customShare
+            ?: throw IllegalStateException("Social+Custom restore way failed. Third share is null")
+        val socialShare = restoreFlowDataLocalRepository.socialShare
+            ?: throw IllegalStateException("Social+Custom restore way failed. Social share is null")
+        val socialShareUserId = restoreFlowDataLocalRepository.socialShareUserId
+            ?: throw IllegalStateException("Social+Custom restore way failed. Social share ID is null")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/restore/UserRestoreInteractor.kt:77]:
```suggestion
        val customShare = restoreFlowDataLocalRepository.customShare
            ?: error("Social+Custom restore way failed. Third share is null")
        val socialShare = restoreFlowDataLocalRepository.socialShare
            ?: error("Social+Custom restore way failed. Social share is null")
        val socialShareUserId = restoreFlowDataLocalRepository.socialShareUserId
            ?: error("Social+Custom restore way failed. Social share ID is null")
```

File: app/src/main/java/org/p2p/wallet/auth/model/OnboardingFlow.kt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package org.p2p.wallet.auth.model
+
+sealed interface OnboardingFlow {
+    object CreateWallet : OnboardingFlow
+
+    open class RestoreWallet : OnboardingFlow {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/model/OnboardingFlow.kt:6]: sealed interfaces are working ok!

Screen Shot 2022-09-15 at 11 28 36

File: app/src/main/java/org/p2p/wallet/auth/model/PhoneNumber.kt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    object PhoneNumberFormatter {
+
+        fun getPhoneNumberE164Formatted(phoneNumber: PhoneNumber): String {
+            val trimmedValue = phoneNumber.formattedValue.replace(" ", "")
+            return "+$trimmedValue"
+        }
+    }
+}
+
+fun PhoneNumber.e164Formatted(): String {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/model/PhoneNumber.kt:18]: move to a method please

**Review by kuznetsov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/infrastructure/InfrastructureModule.kt [hunk 12]
```diff
@@ -76,11 +79,34 @@ object InfrastructureModule : InjectionModule {
             context.getSharedPreferences(name, Context.MODE_PRIVATE)
         }
 
-        single { EncoderDecoderMarshmallow(get()) } bind EncoderDecoder::class
-        single { KeyStore.getInstance(ANDROID_KEY_STORE) }
-        single { KeyStoreWrapper(encoderDecoder = get(), keyStore = get()) }
+        single(named(SHARED_PREFS_ACCOUNT_STORAGE)) {
+            val context = get<Context>()
+            val name = "${context.packageName}.account_prefs"
+            context.getSharedPreferences(name, Context.MODE_PRIVATE)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/InfrastructureModule.kt:85]:
```suggestion
            val name = "${androidContext().packageName}.account_prefs"
            context.getSharedPreferences(name, Context.MODE_PRIVATE)
```

File: app/src/main/java/org/p2p/wallet/infrastructure/InfrastructureModule.kt [hunk 13]
```diff
@@ -36,7 +38,8 @@ import org.p2p.wallet.updates.UpdatesManager
 import java.security.KeyStore
 import java.util.concurrent.Executors
 
-private const val ANDROID_KEY_STORE = "AndroidKeyStore"
+private const val SHARED_PREFS_ACCOUNT_STORAGE = "AccountSharedPreferences"
+private const val SHARED_PREFS_KEY_STORE = "KeyStoreSharedPreferences"
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/InfrastructureModule.kt:42]: please use enum for named injections

File: app/src/main/java/org/p2p/wallet/infrastructure/InfrastructureModule.kt [hunk 14]
```diff
@@ -76,11 +79,34 @@ object InfrastructureModule : InjectionModule {
...
+            val name = "${context.packageName}.account_prefs"
+            context.getSharedPreferences(name, Context.MODE_PRIVATE)
+        }
+        single(named(SHARED_PREFS_KEY_STORE)) {
+            val context = get<Context>()
+            val name = "${context.packageName}.keystore_prefs"
+            context.getSharedPreferences(name, Context.MODE_PRIVATE)
+        }
+
+        single { EncoderDecoderMarshmallow(get(named(SHARED_PREFS_KEY_STORE))) } bind EncoderDecoder::class
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/InfrastructureModule.kt:93]:
```suggestion
        single< EncoderDecoder>{ EncoderDecoderMarshmallow(get(named(SHARED_PREFS_KEY_STORE))) }
```

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/gateway/repository/GatewayServiceErrorMapper.kt (hunk 6 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/auth/gateway/repository/GatewayServiceErrorMapper.kt:34]: I took name error from server )

File: app/src/main/java/org/p2p/wallet/auth/interactor/AuthLogoutInteractor.kt (hunk 7 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/auth/interactor/AuthLogoutInteractor.kt:51]: Oh, this method was duplicate of onUserLogout(), as you can see i clear everything, i just removed useless method

File: app/src/main/java/org/p2p/wallet/auth/model/OnboardingFlow.kt (hunk 10 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/auth/model/OnboardingFlow.kt:6]: i used class, for set Restore without choosing any of flows, cause on first restore step i don't know current user flow

**Pull request merged**